spanner/ot_metrics.go

```diff
@@ -0,0 +1,258 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package spanner
+
+import (
+	"context"
+	"log"
+	"strconv"
+	"strings"
+	"sync"
+
+	"cloud.google.com/go/spanner/internal"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric"
+	"google.golang.org/grpc/metadata"
+)
+
+// OtInstrumentationScope is the instrumentation name that will be associated with the emitted telemetry.
+const OtInstrumentationScope = "cloud.google.com/go"
+const metricsPrefix = "spanner/"
+
+var (
+	attributeKeyClientID   = attribute.Key("client_id")
+	attributeKeyDatabase   = attribute.Key("database")
+	attributeKeyInstance   = attribute.Key("instance_id")
+	attributeKeyLibVersion = attribute.Key("library_version")
+	attributeKeyType       = attribute.Key("type")
+	attributeKeyMethod     = attribute.Key("grpc_client_method")
+
+	attributeNumInUseSessions = attributeKeyType.String("num_in_use_sessions")
+	attributeNumSessions      = attributeKeyType.String("num_sessions")
+	// openTelemetryMetricsEnabled is used to track if OpenTelemetry Metrics need to be recorded
+	openTelemetryMetricsEnabled = false
+	// mutex to avoid data race in reading/writing the above flag
+	otMu = sync.RWMutex{}
+)
+
+func getOpenTelemetryConfig(mp metric.MeterProvider, logger *log.Logger, sessionClientID string, db string) (*openTelemetryConfig, error) {
+	logf(logger, "Inside getOpenTelemetryConfig()")
+	config := &openTelemetryConfig{
+		attributeMap: []attribute.KeyValue{},
+	}
+	if !IsOpenTelemetryMetricsEnabled() {
+		return config, nil
+	}
+	logf(logger, "OT Metrics enbaled in getOpenTelemetryConfig()")
+	_, instance, database, err := parseDatabaseName(db)
+	if err != nil {
+		return nil, err
+	}
+
+	// Construct attributes for Metrics
+	attributeMap := []attribute.KeyValue{
+		attributeKeyClientID.String(sessionClientID),
+		attributeKeyDatabase.String(database),
+		attributeKeyInstance.String(instance),
+		attributeKeyLibVersion.String(internal.Version),
+	}
+	config.attributeMap = append(config.attributeMap, attributeMap...)
+
+	setOpenTelemetryMetricProvider(config, mp, logger)
+	return config, nil
+}
+
+func setOpenTelemetryMetricProvider(config *openTelemetryConfig, mp metric.MeterProvider, logger *log.Logger) {
+	// Fallback to global meter provider in OpenTelemetry
+	if mp == nil {
+		mp = otel.GetMeterProvider()
+	}
+	config.meterProvider = mp
+	initializeMetricInstruments(config, logger)
+}
+
+func initializeMetricInstruments(config *openTelemetryConfig, logger *log.Logger) {
+	if !IsOpenTelemetryMetricsEnabled() {
+		return
+	}
+	meter := config.meterProvider.Meter(OtInstrumentationScope, metric.WithInstrumentationVersion(internal.Version))
+
+	openSessionCountInstrument, err := meter.Int64ObservableGauge(
+		metricsPrefix+"open_session_count1",
+		metric.WithDescription("Number of sessions currently opened"),
+		metric.WithUnit("1"),
```

Comment:
What does `metric.WithUnit("1")` mean? Does that meant that it is an integer?

Reply:
`WithUnit` is used to set the instrument unit.
I followed the same convention used for OpenCensus where we set "1" for [dimensionless](https://github.com/census-instrumentation/opencensus-go/blob/v0.24.0/stats/units.go#L22) unit or count as per [UCCM](https://ucum.org/ucum).

Example in OC:
https://github.com/googleapis/google-cloud-go/blob/1cf8cf1d32c0f84ff8e4fd388ebcce5f5866bed1/spanner/stats.go#L94-L97

Reply:
ack